test/extensions/filters/http/common/fuzz/filter_fuzz_test.cc

```diff
@@ -0,0 +1,122 @@
+#include "test/fuzz/fuzz_runner.h"
+#include "test/fuzz/utility.h"
+#include "test/mocks/buffer/mocks.h"
+#include "test/mocks/http/mocks.h"
+#include "test/mocks/server/mocks.h"
+
+#include "libprotobuf_mutator/src/mutator.h"
+
+namespace Envoy {
+namespace Extensions {
+namespace HttpFilters {
+
+class UberFilterFuzzer {
+public:
+  UberFilterFuzzer() {
+    // Need to set for both a decoder filter and an encoder/decoder filter.
+    ON_CALL(filter_callback_, addStreamDecoderFilter(_))
+        .WillByDefault(
+            Invoke([&](std::shared_ptr<Envoy::Http::StreamDecoderFilter> filter) -> void {
+              filter_ = filter;
+              filter_->setDecoderFilterCallbacks(callbacks_);
+            }));
+    ON_CALL(filter_callback_, addStreamFilter(_))
+        .WillByDefault(
+            Invoke([&](std::shared_ptr<Envoy::Http::StreamDecoderFilter> filter) -> void {
+              filter_ = filter;
+              filter_->setDecoderFilterCallbacks(callbacks_);
+            }));
+    callbacks_.stream_info_.protocol_ = Envoy::Http::Protocol::Http2;
+  }
+
+  // This executes the methods to be fuzzed.
+  void decode(Http::StreamDecoderFilter* filter, const test::fuzz::HttpData& data) {
+    bool end_stream = false;
+
+    Http::TestHeaderMapImpl headers = Fuzz::fromHeaders(data.headers());
+    if (data.data().size() == 0 && !data.has_trailers()) {
+      end_stream = true;
+    }
+    ENVOY_LOG_MISC(debug, "Decoding headers: {} ", data.headers().DebugString());
+    const auto& headersStatus = filter->decodeHeaders(headers, end_stream);
+    if (headersStatus != Http::FilterHeadersStatus::Continue ||
+        headersStatus != Http::FilterHeadersStatus::StopIteration) {
+      return;
+    }
+
+    for (int i = 0; i < data.data().size(); i++) {
+      if (i == data.data().size() - 1 && !data.has_trailers()) {
+        end_stream = true;
+      }
+      Buffer::OwnedImpl buffer(data.data().Get(i));
+      ENVOY_LOG_MISC(debug, "Decoding data: {} ", buffer.toString());
+      if (filter->decodeData(buffer, end_stream) != Http::FilterDataStatus::Continue) {
+      }
+    }
+
+    if (data.has_trailers()) {
+      ENVOY_LOG_MISC(debug, "Decoding trailers: {} ", data.trailers().DebugString());
+      Http::TestHeaderMapImpl trailers = Fuzz::fromHeaders(data.trailers());
+      filter->decodeTrailers(trailers);
+    }
+  }
+
+  // This creates and mutates the filter config and runs decode.
+  void fuzz(absl::string_view filter_name, const test::fuzz::HttpData& data) {
+    ENVOY_LOG_MISC(info, "Fuzzing filter: {}", filter_name);
+
+    auto& factory =
+        Config::Utility::getAndCheckFactory<Server::Configuration::NamedHttpFilterConfigFactory>(
+            std::string(filter_name));
+    auto proto_config = factory.createEmptyConfigProto();
+
+    protobuf_mutator::Mutator mutator;
+    try {
+      // Mutate and validate the chosen filter protobuf directly.
+      mutator.Mutate(proto_config.get(), 200);
```

Comment:
I think we're just running the mutator here directly on an empty config. I wonder if instead we could make use of something in the input proto, so that the mutation is now a function of the selected filter's config type _and_ some input entropy. 

In an ideal world, there would be some type magic so that the outer LPM, driving the fuzz test, had some oracle to know the proto type that was going to be selected fro the test run. This magic doesn't exist, so I think the next best thing is for the outer LPM to produce some bit vector that we combine with the known proto type here with an inner LPM.

Maybe there is a smarter way to do this that allows the outer LPM to be the only LPM and/or allows it to do smarter cross-over and mutations, but I'm not seeing it.

Reply:
Instead of fuzzing a filter_index and mutating an empty config, I opted for fuzzing configs, and registering a post processor mutation that will always ensure that we at least consume valid named configs. 

It doesn't solve the problem of the typed_config containing an Any which may or may not be parsed as the right type, or that the text proto representation of the Any is binary encoded and ugly.

Reply:
This makes sense, thanks. I think there are two things to think about here:
1. Agree on the corpus and generation, we could get a good baseline by generating this from unit tests. I think we would want a framework that makes this basically automatic any time you write a unit test for an HTTP filter.
2. The mutation of the Any seems to me to be an area that is still a bit hard to understand. LPM is going to see a sequence of bytes and do byte mutation rather than structured proto mutation I think. Could you tag the LPM author on this thread? Would be good to get their take on this.

Reply:
@vitalybuka Would love some feedback on this fuzzer. We have a proto fuzzer that takes as input a message that has an embedded `google.protobuf.Any`. https://github.com/envoyproxy/envoy/blob/c3bddaee1912fcd1fedc4786aee830b2e4a7c599/api/envoy/extensions/filters/network/http_connection_manager/v3alpha/http_connection_manager.proto#L640

* The PostProcessorRegistration constrains the filter name to match a supported filter. 
* The type of the `typed_config` field depends on the filter being instantiated. Because it is an `Any`, the value is a serialized message. Assuming LPM will perform byte mutation on this field, is there a better way to perform structured fuzzing on the value?

Reply:
Update: I added some valid typed_configs to represent the buffer_filter into a corpus directory, and some `ASSERT(false)` in the source code to check if we ever generate a config with certain values for fields, and running the fuzzer hundreds of thousands of runs doesn't catch the assert. I logged the config we produce when fuzzing, and too many aren't valid.

@htuch specifically, I crashed if we ever set max request bytes to 0, and restricted to only fuzzer buffer filter. Too many inputs turned into `hidden_envoy_deprecated_config` with deeper and deeper struct definitions.